### PR TITLE
feat: add `BitVec.(getMsbD, getLsbD, msb)_(neg, abs)`

### PR DESCRIPTION
**Pull request by luisacicolini**

### CODE REVIEW

File: SSA/Projects/InstCombine/ForLean.lean [hunk 1]
```diff
@@ -559,10 +550,29 @@ theorem msb_sub {x y: BitVec w} :
       = (x.getMsbD 0 ^^ ((~~~y + 1#w).getMsbD 0 ^^ carry (w - 1 - 0) x (~~~y + 1#w) false)) := by
   simp [sub_eq_add_neg, BitVec.neg_eq_not_add, msb_add]
 
-theorem msb_neg {w : Nat} {x : BitVec w} (h : 0 < w) :
-    (~~~x).msb = (getMsbD x 0).not := by
-  rw [BitVec.msb, BitVec.getMsbD_neg]
-  omega
+theorem msb_neg  {x : BitVec w} :
+    (-x).msb = (~~~x + 1#w).msb := by
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem msb_neg {x : BitVec w} :
    (-x).msb = (~~~x + 1#w).msb := by
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 2]
```diff
@@ -559,10 +550,29 @@ theorem msb_sub {x y: BitVec w} :
...
+  rw [neg_eq_not_add]
+
+theorem getLsbD_neg {i : Nat}  {x : BitVec w} :
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem getLsbD_neg {i : Nat} {x : BitVec w} :
```

**Review by tobiasgrosser — APPROVED** [review]

File: SSA/Projects/InstCombine/ForLean.lean [hunk 3]
```diff
@@ -559,10 +550,29 @@ theorem msb_sub {x y: BitVec w} :
...
+    x.abs.getLsbD i = (if x.msb = true then -x else x).getLsbD i := by
+  simp [BitVec.abs]
+
+theorem getMsbD_abs {i : Nat} {x : BitVec w} :
+    getMsbD (x.abs) i = getMsbD (if x.msb = true then -x else x) i := by
+  simp [BitVec.abs]
+
+theorem msb_abs {w : Nat} {x : BitVec w} :
+    (x.abs).msb = (if x.msb = true then -x else x).msb := by
+  simp [BitVec.abs]
```
Comment (alexkeizer) [inline SSA/Projects/InstCombine/ForLean.lean]: You might want to be a bit more consistent with wether or not you use dot-notation here, in particular, `(if x.msb = true then -x else x).getLsbD` vs `getMsbD (if x.msb = true then -x else x)`

**Comment by luisacicolini** [comment]
thanks for your feedback @tobiasgrosser @alexkeizer. Do you think that this is in a state where it can be upstreamed to lean4 already?

**Comment by tobiasgrosser** [comment]
I think so.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 4]
```diff
@@ -559,10 +550,29 @@ theorem msb_sub {x y: BitVec w} :
...
-    (~~~x).msb = (getMsbD x 0).not := by
-  rw [BitVec.msb, BitVec.getMsbD_neg]
-  omega
+theorem msb_neg {x : BitVec w} :
+    (-x).msb = (~~~x + 1#w).msb := by
+  rw [neg_eq_not_add]
+
+theorem getLsbD_neg {i : Nat} {x : BitVec w} :
+    getLsbD (-x) i = getLsbD (~~~x + 1#w) i := by
+  simp [neg_eq_not_add]
```
Comment (luisacicolini) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
  rw [neg_eq_not_add]
```

**Pull request merged**